File: apps/web/lib/api/jobs.ts
import { getApiBaseUrl } from "./config";

/** M0e + A3 CreateJobRequest (wire camelCase). */
export type CreateJobRequest = {
  visionText: string;
  inspirationAssetIds?: string[];
  clientMetadata?: Record<string, unknown>;
  /** OpenRouter model id from Create picker (must be server-allowed). */
  model?: string;
  /** A3: run clarify interview before plan/codegen. */
  planMode?: boolean;
};

export type JobStatus =
  | "queued"
  | "running"
  | "awaiting_clarify"
  | "succeeded"
  | "failed";

export type JobPhase = "clarify" | "plan" | "codegen" | "validate" | "repair";

export type QuotaFields = {
  createsUsed: number;
  createsLimit: number;
  resetsAt?: string;
};

export type ClarifyOption = {
  value: string;
  label: string;
  description?: string;
};

export type ClarifyQuestion = {
  id: string;
  prompt: string;
  options: ClarifyOption[];
  multiSelect?: boolean;
  allowAllOptions?: boolean;
  group?: string;
};

export type ClarifyAnswerValue =
  | string
  | string[]
  | { type: "all_options" };

export type ClarifyForcedEnum = {
  name: string;
  label: string;
  options: Array<{ value: string; label: string }>;
  default: string;
  group?: string;
  sourceQuestionId: string;
};

export type ClarifyResult = {
  transcript: string;
  forcedEnums: ClarifyForcedEnum[];
  lockedNotes: string[];
  summary?: string;
};

export type JobClarifyState = {
  understanding?: string;
  questions?: ClarifyQuestion[];
  answers?: Record<string, ClarifyAnswerValue>;
  result?: ClarifyResult;
  answered?: boolean;
  skipReason?: string;
};

/** M0e CreateJobResponse (+ optional debug userId). */
export type CreateJobResponse = {
  jobId: string;
  status: JobStatus;
  createdAt: string;
  userId?: string;
  quota?: QuotaFields;
  planMode?: boolean;
};

/** Persisted chat turn on a generation job. */
export type JobChatMessage = {
  id: string;
  role: "user" | "assistant";
  content: string;
  kind?: string | null;
  createdAt: string;
  meta?: Record<string, unknown> | null;
};

export type JobStatusResponse = {
  jobId: string;
  status: JobStatus;
  phase?: JobPhase | null;
  progress?: number | null;
  errorCode?: string | null;
  errorMessage?: string | null;
  resultReady?: boolean | null;
  updatedAt?: string | null;
  repair?: {
    maxRepairs: number;
    repairsUsed: number;
    tokenBudget?: number | null;
    tokensUsed?: number | null;
    wallTimeMs?: number | null;
    wallTimeUsedMs?: number | null;
  } | null;
  quota?: QuotaFields | null;
  planMode?: boolean | null;
  clarify?: JobClarifyState | null;
  /** Ordered user/assistant history for this job. */
  messages?: JobChatMessage[] | null;
};

export type ClarifyJobRequest = {
  answers: Record<string, ClarifyAnswerValue>;
  buildNow?: boolean;
};

export type ClarifyJobResponse = {
  jobId: string;
  status: JobStatus;
  clarify?: JobClarifyState | null;
  updatedAt?: string | null;
};

export type JobResultResponse = {
  jobId: string;
  toolId: string;
  versionId: string;
  target: string;
  publicId?: string | null;
  completedAt?: string | null;
};

async function readError(res: Response): Promise<string> {
  const text = await res.text().catch(() => "");
  return text ? `: ${text}` : "";
}

export class CreateJobApiError extends Error {
  readonly status: number;
  readonly errorCode?: string;
  readonly quota?: QuotaFields;

  constructor(
    message: string,
    opts: { status: number; errorCode?: string; quota?: QuotaFields },
  ) {
    super(message);
    this.name = "CreateJobApiError";
    this.status = opts.status;
    this.errorCode = opts.errorCode;
    this.quota = opts.quota;
  }
}

/**
 * POST /api/v1/jobs with session cookie (credentials include).
 * M3a+: persists job + draft tool; may 429 on quota.
 */
export async function createJob(
  body: CreateJobRequest,
): Promise<CreateJobResponse> {
  const res = await fetch(`${getApiBaseUrl()}/api/v1/jobs`, {
    method: "POST",
    credentials: "include",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    let errorCode: string | undefined;
    let quota: QuotaFields | undefined;
    let message = `Create job failed (${res.status})${text ? `: ${text}` : ""}`;
    try {
      const parsed = JSON.parse(text) as {
        errorCode?: string;
        errorMessage?: string;
        quota?: QuotaFields;
      };
      if (parsed.errorCode) errorCode = parsed.errorCode;
      if (parsed.quota) quota = parsed.quota;
      if (parsed.errorMessage) {
        message = parsed.errorMessage;
      }
    } catch {
      /* plain text body */
    }
    throw new CreateJobApiError(message, {
      status: res.status,
      errorCode,
      quota,
    });
  }

  return res.json() as Promise<CreateJobResponse>;
}

/** POST /api/v1/jobs/{jobId}/clarify — A3 answers → resume build. */
export async function submitClarify(
  jobId: string,
  body: ClarifyJobRequest,
): Promise<ClarifyJobResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/jobs/${encodeURIComponent(jobId)}/clarify`,
    {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    },
  );

  if (!res.ok) {
    const text = await res.text().catch(() => "");
    let message = `Clarify failed (${res.status})${text ? `: ${text}` : ""}`;
    try {
      const parsed = JSON.parse(text) as {
        detail?: string;
        errorMessage?: string;
      };
      if (parsed.errorMessage) message = parsed.errorMessage;
      else if (typeof parsed.detail === "string") message = parsed.detail;
    } catch {
      /* plain */
    }
    throw new CreateJobApiError(message, { status: res.status });
  }

  return res.json() as Promise<ClarifyJobResponse>;
}

/** Parse salvage tool id from failed job errorMessage (M3e). */
export function parseSalvageToolId(
  errorMessage: string | null | undefined,
): string | null {
  if (!errorMessage) return null;
  const m = errorMessage.match(/salvage_draft=true\s+toolId=([0-9a-fA-F-]{36})/);
  return m?.[1] ?? null;
}

/** GET /api/v1/jobs/{jobId} — owner poll. */
export async function getJobStatus(jobId: string): Promise<JobStatusResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/jobs/${encodeURIComponent(jobId)}`,
    {
      method: "GET",
      credentials: "include",
    },
  );

  if (!res.ok) {
    throw new Error(
      `Get job status failed (${res.status})${await readError(res)}`,
    );
  }

  return res.json() as Promise<JobStatusResponse>;
}

/**
 * GET /api/v1/jobs/{jobId}/result — only when status is succeeded.
 * Returns 409 while still queued/running/failed.
 */
export async function getJobResult(jobId: string): Promise<JobResultResponse> {
  const res = await fetch(
    `${getApiBaseUrl()}/api/v1/jobs/${encodeURIComponent(jobId)}/result`,
    {
      method: "GET",
      credentials: "include",
    },
  );

  if (!res.ok) {
    throw new Error(
      `Get job result failed (${res.status})${await readError(res)}`,
    );
  }

  return res.json() as Promise<JobResultResponse>;
}

File: apps/api/src/schemas/jobs.py
"""
Job API shapes (M0e / M1a).

Wire format is camelCase to match `@repo/contracts` / md/contracts/job-api.md.
Python field names use snake_case with aliases.
"""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

JobStatus = Literal[
    "queued",
    "running",
    "awaiting_clarify",
    "succeeded",
    "failed",
]
JobPhase = Literal["clarify", "plan", "codegen", "validate", "repair"]
JobErrorCode = Literal[
    "UNAUTHORIZED",
    "QUOTA_EXCEEDED",
    "VALIDATION_FAILED",
    "GENERATION_FAILED",
    "TIMEOUT",
    "INTERNAL",
]

class CamelModel(BaseModel):
    """Base model: accept snake_case or camelCase, serialize camelCase."""

    model_config = ConfigDict(
        populate_by_name=True,
        serialize_by_alias=True,
        extra="ignore",
    )


class QuotaFields(CamelModel):
    creates_used: int = Field(alias="createsUsed")
    creates_limit: int = Field(alias="createsLimit")
    resets_at: str | None = Field(default=None, alias="resetsAt")


class RepairBudgetFields(CamelModel):
    max_repairs: int = Field(alias="maxRepairs")
    repairs_used: int = Field(alias="repairsUsed")
    token_budget: int | None = Field(default=None, alias="tokenBudget")
    tokens_used: int | None = Field(default=None, alias="tokensUsed")
    wall_time_ms: int | None = Field(default=None, alias="wallTimeMs")
    wall_time_used_ms: int | None = Field(default=None, alias="wallTimeUsedMs")


class CreateJobRequest(CamelModel):
    """POST /api/v1/jobs body."""

    vision_text: str = Field(alias="visionText", min_length=1)
    inspiration_asset_ids: list[str] | None = Field(
        default=None,
        alias="inspirationAssetIds",
    )
    client_metadata: dict[str, Any] | None = Field(
        default=None,
        alias="clientMetadata",
    )
    # Optional OpenRouter model id from Create picker (must be in LLM_MODELS_ALLOWED).
    model: str | None = None
    # A3: opt-in clarify interview before plan/codegen.
    plan_mode: bool = Field(default=False, alias="planMode")


class CreateJobResponse(CamelModel):
    """POST /api/v1/jobs accept response (usually status queued)."""

    job_id: str = Field(alias="jobId")
    status: JobStatus
    created_at: str = Field(alias="createdAt")
    quota: QuotaFields | None = None
    plan_mode: bool | None = Field(default=None, alias="planMode")
    # Stub-only debug aid (not in TS CreateJobResponse). Remove before external beta.
    user_id: str | None = Field(default=None, alias="userId")


class ClarifyJobRequest(CamelModel):
    """POST /api/v1/jobs/{jobId}/clarify body (A3)."""

    answers: dict[str, Any]
    build_now: bool = Field(default=True, alias="buildNow")


class ClarifyJobResponse(CamelModel):
    """POST clarify accept — usually re-queued for build."""

    job_id: str = Field(alias="jobId")
    status: JobStatus
    clarify: dict[str, Any] | None = None
    updated_at: str | None = Field(default=None, alias="updatedAt")


class RefineJobRequest(CamelModel):
    """POST /api/v1/tools/{toolId}/refine body (AM7)."""

    message: str = Field(min_length=1)
    base_version_id: str | None = Field(default=None, alias="baseVersionId")


class RefineBudgetFields(CamelModel):
    refine_used: int = Field(alias="refineUsed")
    refine_limit: int = Field(alias="refineLimit")


class RefineJobResponse(CamelModel):
    """POST refine accept response."""

    job_id: str = Field(alias="jobId")
    tool_id: str = Field(alias="toolId")
    base_version_id: str = Field(alias="baseVersionId")
    status: JobStatus
    created_at: str = Field(alias="createdAt")
    job_kind: str = Field(default="refine", alias="jobKind")
    refine: RefineBudgetFields | None = None


class JobStatusResponse(CamelModel):
    """GET /api/v1/jobs/:jobId poll response (M3 + A3)."""

    job_id: str = Field(alias="jobId")
    status: JobStatus
    phase: JobPhase | None = None
    progress: float | None = None
    error_code: JobErrorCode | None = Field(default=None, alias="errorCode")
    error_message: str | None = Field(default=None, alias="errorMessage")
    quota: QuotaFields | None = None
    repair: RepairBudgetFields | None = None
    updated_at: str | None = Field(default=None, alias="updatedAt")
    result_ready: bool | None = Field(default=None, alias="resultReady")
    plan_mode: bool | None = Field(default=None, alias="planMode")
    clarify: dict[str, Any] | None = None


class JobResultResponse(CamelModel):
    """GET /api/v1/jobs/:jobId/result success payload (M3)."""

    job_id: str = Field(alias="jobId")
    tool_id: str = Field(alias="toolId")
    version_id: str = Field(alias="versionId")
    target: str
    public_id: str | None = Field(default=None, alias="publicId")
    completed_at: str | None = Field(default=None, alias="completedAt")


class JobErrorBody(CamelModel):
    error_code: JobErrorCode = Field(alias="errorCode")
    error_message: str = Field(alias="errorMessage")
    job_id: str | None = Field(default=None, alias="jobId")
    quota: QuotaFields | None = None
